apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/internal/SortedInputFieldMapWriter.kt

```diff
@@ -0,0 +1,148 @@
+  }
+
+  override fun writeBoolean(fieldName: String, value: Boolean?) {
+    buffer[fieldName] = value
+  }
+
+  override fun writeCustom(fieldName: String, scalarType: ScalarType, value: Any?) {
+    buffer[fieldName] = value
+  }
+
+  @Throws(IOException::class)
+  override fun writeObject(fieldName: String, marshaller: InputFieldMarshaller?) {
+    if (marshaller == null) {
+      buffer[fieldName] = null
+    } else {
+      val nestedWriter = SortedInputFieldMapWriter(fieldNameComparator)
+      marshaller.marshal(nestedWriter)
+      buffer[fieldName] = nestedWriter.buffer
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun writeList(fieldName: String, listWriter: InputFieldWriter.ListWriter?) {
+    if (listWriter == null) {
+      buffer[fieldName] = null
+    } else {
+      val listItemWriter = ListItemWriter(fieldNameComparator)
+      listWriter.write(listItemWriter)
+      buffer[fieldName] = listItemWriter.list
+    }
+  }
+
+  override fun writeMap(fieldName: String, value: Map<String, Any?>?) {
+    buffer[fieldName] = value
+  }
+
+  private open class ListItemWriter internal constructor(val fieldNameComparator: Comparator<String>) : InputFieldWriter.ListItemWriter {
+    val list: MutableList<Any?> = ArrayList<Any?>()
+    override fun writeString(value: String?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    override fun writeInt(value: Int?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    override fun writeLong(value: Long?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    override fun writeDouble(value: Double?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    override fun writeNumber(value: Number?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    override fun writeBoolean(value: Boolean?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    override fun writeCustom(scalarType: ScalarType, value: Any?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+    @Throws(IOException::class)
+    override fun writeObject(marshaller: InputFieldMarshaller?) {
+      if (marshaller != null) {
+        val nestedWriter = SortedInputFieldMapWriter(fieldNameComparator)
+        marshaller.marshal(nestedWriter)
+        list.add(nestedWriter.buffer)
+      }
+    }
+
+    @Throws(IOException::class)
+    override fun writeList(listWriter: InputFieldWriter.ListWriter?) {
+      if (listWriter != null) {
+        val nestedListItemWriter = ListItemWriter(fieldNameComparator)
+        listWriter.write(nestedListItemWriter)
+        list.add(nestedListItemWriter.list)
+      }
+    }
+
+    override fun writeMap(value: Map<String, Any?>?) {
+      if (value != null) {
+        list.add(value)
+      }
+    }
+
+  }
+
+  init {
+    this.fieldNameComparator = __checkNotNull(fieldNameComparator, "fieldNameComparator == null")
```

Comment:
Instead of `init`, can you move this to constructor. We also shouldn't use `checkNotNull` in kotlin.

Reply:
Sorry I forgot to check the automatic code conversion there 😅. Thank for the heads up!